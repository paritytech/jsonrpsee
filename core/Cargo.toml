[package]
name = "jsonrpsee-core"
version = "0.6.1"
authors = ["Parity Technologies <admin@parity.io>"]
description = "Utilities for jsonrpsee"
edition = "2021"
license = "MIT"

[dependencies]
anyhow = "1"
arrayvec = "0.7.1"
async-trait = "0.1"
beef = { version = "0.5.1", features = ["impl_serde"] }
thiserror = "1"
futures-channel = { version = "0.3.14", default-features = false }
futures-util = { version = "0.3.14", default-features = false, optional = true }
hyper = { version = "0.14.10", default-features = false, features = ["stream"] }
jsonrpsee-types = { path = "../types", version = "0.6.1" }
tracing = { version = "0.1", optional = true }
rustc-hash = { version = "1", optional = true }
rand = { version = "0.8", optional = true }
serde = { version = "1.0", default-features = false, features = ["derive"] }
serde_json = { version = "1", features = ["raw_value"] }
soketto = "0.7.1"
parking_lot = { version = "0.11", optional = true }
tokio = { version = "1.8", features = ["rt"], optional = true }

[features]
default = []
http-helpers = ["futures-util"]
server = [
	"futures-util",
	"rustc-hash",
	"tracing",
	"parking_lot",
	"rand",
	"tokio",
]
client = ["futures-util"]
async-client = [
	"client",
	"rustc-hash",
	"tokio/sync",
	"tokio/macros",
	"tokio/time",
	"tracing"
]

[dev-dependencies]
serde_json = "1.0"
tokio = { version = "1.8", features = ["macros", "rt"] }
jsonrpsee = { path = "../jsonrpsee", features = ["server", "macros"] }
