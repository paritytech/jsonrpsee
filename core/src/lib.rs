// Copyright 2019-2021 Parity Technologies (UK) Ltd.
//
// Permission is hereby granted, free of charge, to any
// person obtaining a copy of this software and associated
// documentation files (the "Software"), to deal in the
// Software without restriction, including without
// limitation the rights to use, copy, modify, merge,
// publish, distribute, sublicense, and/or sell copies of
// the Software, and to permit persons to whom the Software
// is furnished to do so, subject to the following
// conditions:
//
// The above copyright notice and this permission notice
// shall be included in all copies or substantial portions
// of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF
// ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED
// TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
// PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT
// SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
// CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR
// IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
// DEALINGS IN THE SOFTWARE.

//! Shared utilities for `jsonrpsee`.

#![warn(missing_docs, missing_debug_implementations, unreachable_pub)]

/// Error type.
pub mod error;

/// Traits
pub mod traits;

/// Middleware trait and implementation.
pub mod middleware;

/// Shared hyper helpers.
#[cfg(feature = "http-helpers")]
pub mod http_helpers;

/// Shared code for JSON-RPC servers.
#[cfg(feature = "server")]
pub mod server;

/// Shared code for JSON-RPC clients.
#[cfg(feature = "client")]
pub mod client;

pub use async_trait::async_trait;
pub use error::Error;

/// JSON-RPC result.
pub type RpcResult<T> = std::result::Result<T, Error>;

/// Re-exports for proc-macro library to not require any additional
/// dependencies to be explicitly added on the client side.
#[doc(hidden)]
pub mod __reexports {
	pub use async_trait::async_trait;
	pub use serde;
	pub use serde_json;
}

pub use beef::Cow;
pub use serde::{de::DeserializeOwned, Serialize};
pub use serde_json::{
	to_value as to_json_value, value::to_raw_value as to_json_raw_value, value::RawValue as JsonRawValue,
	Value as JsonValue,
};

/// Ten megabytes.
pub const TEN_MB_SIZE_BYTES: u32 = 10 * 1024 * 1024;
