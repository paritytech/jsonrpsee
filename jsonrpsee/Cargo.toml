[package]
name = "jsonrpsee"
description = "JSON-RPC crate"
version = "0.2.0"
authors = ["Parity Technologies <admin@parity.io>", "Pierre Krieger <pierre.krieger1708@gmail.com>"]
license = "MIT"
edition = "2018"
repository = "https://github.com/paritytech/jsonrpsee"
homepage = "https://github.com/paritytech/jsonrpsee"
documentation = "https://docs.rs/jsonrpsee"

[dependencies]
http-client = { path = "../http-client", version = "0.2.0", package = "jsonrpsee-http-client", optional = true }
http-server = { path = "../http-server", version = "0.2.0", package = "jsonrpsee-http-server", optional = true }
ws-client = { path = "../ws-client", version = "0.2.0", package = "jsonrpsee-ws-client", optional = true }
ws-server = { path = "../ws-server", version = "0.2.0", package = "jsonrpsee-ws-server", optional = true }
proc-macros = { path = "../proc-macros", version = "0.2.0", package = "jsonrpsee-proc-macros", optional = true }
utils = { path = "../utils", version = "0.2.0", package = "jsonrpsee-utils", optional = true }
types = { path = "../types", version = "0.2.0", package = "jsonrpsee-types", optional = true }

[features]
client = ["http-client", "ws-client"]
server = ["http-server", "ws-server", "utils"]
macros = ["proc-macros", "types"]
full = ["client", "server", "macros", "utils"]
