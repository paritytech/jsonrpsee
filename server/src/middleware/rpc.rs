// Copyright 2019-2021 Parity Technologies (UK) Ltd.
//
// Permission is hereby granted, free of charge, to any
// person obtaining a copy of this software and associated
// documentation files (the "Software"), to deal in the
// Software without restriction, including without
// limitation the rights to use, copy, modify, merge,
// publish, distribute, sublicense, and/or sell copies of
// the Software, and to permit persons to whom the Software
// is furnished to do so, subject to the following
// conditions:
//
// The above copyright notice and this permission notice
// shall be included in all copies or substantial portions
// of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF
// ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED
// TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
// PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT
// SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
// CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR
// IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
// DEALINGS IN THE SOFTWARE.

//! JSON-RPC service middleware.

pub use jsonrpsee_core::middleware::*;
pub use jsonrpsee_core::server::MethodResponse;

use std::sync::Arc;

use crate::ConnectionId;
use jsonrpsee_core::server::{
	BatchResponseBuilder, BoundedSubscriptions, MethodCallback, MethodSink, Methods, SubscriptionState,
};
use jsonrpsee_core::traits::IdProvider;
use jsonrpsee_types::ErrorObject;
use jsonrpsee_types::error::{ErrorCode, reject_too_many_subscriptions};

/// JSON-RPC service middleware.
#[derive(Clone, Debug)]
pub struct RpcService {
	conn_id: ConnectionId,
	methods: Methods,
	max_response_body_size: usize,
	cfg: RpcServiceCfg,
}

/// Configuration of the RpcService.
#[derive(Clone, Debug)]
pub(crate) enum RpcServiceCfg {
	/// The server supports only calls.
	OnlyCalls,
	/// The server supports both method calls and subscriptions.
	CallsAndSubscriptions {
		bounded_subscriptions: BoundedSubscriptions,
		sink: MethodSink,
		id_provider: Arc<dyn IdProvider>,
		_pending_calls: tokio::sync::mpsc::Sender<()>,
	},
}

impl RpcService {
	/// Create a new service.
	pub(crate) fn new(
		methods: Methods,
		max_response_body_size: usize,
		conn_id: ConnectionId,
		cfg: RpcServiceCfg,
	) -> Self {
		Self { methods, max_response_body_size, conn_id, cfg }
	}
}

impl RpcServiceT for RpcService {
	type BatchResponse = MethodResponse;
	type MethodResponse = MethodResponse;
	type NotificationResponse = MethodResponse;

	fn call<'a>(&self, req: Request<'a>) -> impl Future<Output = Self::MethodResponse> + Send + 'a {
		let conn_id = self.conn_id;
		let max_response_body_size = self.max_response_body_size;

		let Request { id, method, params, extensions, .. } = req;
		let params = jsonrpsee_types::Params::new(params.as_ref().map(|p| serde_json::value::RawValue::get(p)));

		match self.methods.method_with_name(&method) {
			None => {
				let rp =
					MethodResponse::error(id, ErrorObject::from(ErrorCode::MethodNotFound)).with_extensions(extensions);
				ResponseFuture::ready(rp)
			}
			Some((_name, method)) => match method {
				MethodCallback::Async(callback) => {
					let params = params.into_owned();
					let id = id.into_owned();
					let fut = (callback)(id, params, conn_id, max_response_body_size, extensions);

					ResponseFuture::future(fut)
				}
				MethodCallback::Sync(callback) => {
					let rp = (callback)(id, params, max_response_body_size, extensions);
					ResponseFuture::ready(rp)
				}
				MethodCallback::Subscription(callback) => {
					let RpcServiceCfg::CallsAndSubscriptions {
						bounded_subscriptions,
						sink,
						id_provider,
						_pending_calls,
					} = self.cfg.clone()
					else {
						tracing::warn!("Subscriptions not supported");
						let rp = MethodResponse::error(id, ErrorObject::from(ErrorCode::InternalError))
							.with_extensions(extensions);
						return ResponseFuture::ready(rp);
					};

					if let Some(p) = bounded_subscriptions.acquire() {
						let conn_state =
							SubscriptionState { conn_id, id_provider: &*id_provider.clone(), subscription_permit: p };

						let fut = (callback)(id.clone(), params, sink, conn_state, extensions);
						ResponseFuture::future(fut)
					} else {
						let max = bounded_subscriptions.max();
						let rp =
							MethodResponse::error(id, reject_too_many_subscriptions(max)).with_extensions(extensions);
						ResponseFuture::ready(rp)
					}
				}
				MethodCallback::Unsubscription(callback) => {
					// Don't adhere to any resource or subscription limits; always let unsubscribing happen!

					let RpcServiceCfg::CallsAndSubscriptions { .. } = self.cfg else {
						tracing::warn!("Subscriptions not supported");
						let rp = MethodResponse::error(id, ErrorObject::from(ErrorCode::InternalError))
							.with_extensions(extensions);
						return ResponseFuture::ready(rp);
					};

					let rp = callback(id, params, conn_id, max_response_body_size, extensions);
					ResponseFuture::ready(rp)
				}
			},
		}
	}

	fn batch<'a>(&self, batch: Batch<'a>) -> impl Future<Output = Self::BatchResponse> + Send + 'a {
		let mut batch_rp = BatchResponseBuilder::new_with_limit(self.max_response_body_size);
		let service = self.clone();
		async move {
			let mut got_notification = false;

			for batch_entry in batch.into_iter() {
				match batch_entry {
					Ok(BatchEntry::Call(req)) => {
						let rp = service.call(req).await;
						if let Err(err) = batch_rp.append(rp) {
							return err;
						}
					}
					Ok(BatchEntry::Notification(n)) => {
						got_notification = true;
						service.notification(n).await;
					}
					Err(err) => {
						let (err, id) = err.into_parts();
						let rp = MethodResponse::error(id, err);
						if let Err(err) = batch_rp.append(rp) {
							return err;
						}
					}
				}
			}

			// If the batch is empty and we got a notification, we return an empty response.
			if batch_rp.is_empty() && got_notification {
				MethodResponse::notification()
			}
			// An empty batch is regarded as an invalid request here.
			else {
				MethodResponse::from_batch(batch_rp.finish())
			}
		}
	}

	fn notification<'a>(&self, n: Notification<'a>) -> impl Future<Output = Self::BatchResponse> + Send + 'a {
		// The notification should not be replied to with a response
		// but we propogate the extensions to the response which can be useful
		// for example HTTP transport to set the headers.
		async move { MethodResponse::notification().with_extensions(n.extensions) }
	}
}
