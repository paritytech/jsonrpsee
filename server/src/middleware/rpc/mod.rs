// Copyright 2019-2021 Parity Technologies (UK) Ltd.
//
// Permission is hereby granted, free of charge, to any
// person obtaining a copy of this software and associated
// documentation files (the "Software"), to deal in the
// Software without restriction, including without
// limitation the rights to use, copy, modify, merge,
// publish, distribute, sublicense, and/or sell copies of
// the Software, and to permit persons to whom the Software
// is furnished to do so, subject to the following
// conditions:
//
// The above copyright notice and this permission notice
// shall be included in all copies or substantial portions
// of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF
// ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED
// TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
// PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT
// SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
// CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR
// IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
// DEALINGS IN THE SOFTWARE.

//! Various middleware implementations for JSON-RPC specific purposes.

mod layers;

pub use layers::*;

use tower::layer::util::{Identity, Stack};
use tower::layer::LayerFn;
use tower::util::Either;

use jsonrpsee_core::server::MethodResponse;
use jsonrpsee_types::Request;

/// Similar to `tower::Service` but specific for jsonrpsee and
/// doesn't requires `&mut self` for performance reasons.
#[async_trait::async_trait]
pub trait RpcServiceT<'a> {
	/// Process a single JSON-RPC call it may be a subscription or regular call.
	/// In this interface they are treated in the same way but it's possible to
	/// distinguish those based on the `MethodResponse`.
	async fn call(&self, request: Request<'a>, transport_label: TransportProtocol) -> MethodResponse;
}

/// The transport protocol used to send or receive a call or request.
#[derive(Debug, Copy, Clone)]
pub enum TransportProtocol {
	/// HTTP transport.
	Http,
	/// WebSocket transport.
	WebSocket,
}

impl std::fmt::Display for TransportProtocol {
	fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
		let s = match self {
			Self::Http => "http",
			Self::WebSocket => "websocket",
		};

		write!(f, "{s}")
	}
}

/// Similar to [`tower::ServiceBuilder`] but doesn't
/// support any tower middleware implementations.
#[derive(Debug, Clone)]
pub struct RpcServiceBuilder<L>(tower::ServiceBuilder<L>);

impl Default for RpcServiceBuilder<Identity> {
	fn default() -> Self {
		RpcServiceBuilder(tower::ServiceBuilder::new())
	}
}

impl RpcServiceBuilder<Identity> {
	/// Create a new [`RpcServiceBuilder`].
	pub fn new() -> Self {
		Self(tower::ServiceBuilder::new())
	}
}

impl<L> RpcServiceBuilder<L> {
	/// Optionally add a new layer `T` to the [`RpcServiceBuilder`].
	///
	/// See the documentation for [`tower::ServiceBuilder::option_layer`] for more details.
	pub fn option_layer<T>(self, layer: Option<T>) -> RpcServiceBuilder<Stack<Either<T, Identity>, L>> {
		RpcServiceBuilder(self.0.option_layer(layer))
	}

	/// Add a new layer `T` to the [`RpcServiceBuilder`].
	///
	/// See the documentation for [`tower::ServiceBuilder::layer`] for more details.
	pub fn layer<T>(self, layer: T) -> RpcServiceBuilder<Stack<T, L>> {
		RpcServiceBuilder(self.0.layer(layer))
	}

	/// Add a [`tower::Layer`] built from a function that accepts a service and returns another service.
	///
	/// See the documentation for [`tower::ServiceBuilder::layer_fn`] for more details.
	pub fn layer_fn<F>(self, f: F) -> RpcServiceBuilder<Stack<LayerFn<F>, L>> {
		RpcServiceBuilder(self.0.layer_fn(f))
	}

	/// Add a logging layer to [`RpcServiceBuilder`]
	///
	/// This logs each request and response for every call.
	///
	pub fn rpc_logger(self, max_log_len: u32) -> RpcServiceBuilder<Stack<RpcLoggerLayer, L>> {
		RpcServiceBuilder(self.0.layer(RpcLoggerLayer::new(max_log_len)))
	}

	/// Wrap the service `S` with the middleware.
	pub(crate) fn service<S>(&self, service: S) -> L::Service
	where
		L: tower::Layer<S>,
	{
		self.0.service(service)
	}
}
