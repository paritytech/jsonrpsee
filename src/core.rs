// Copyright 2019 Parity Technologies (UK) Ltd.
//
// Permission is hereby granted, free of charge, to any
// person obtaining a copy of this software and associated
// documentation files (the "Software"), to deal in the
// Software without restriction, including without
// limitation the rights to use, copy, modify, merge,
// publish, distribute, sublicense, and/or sell copies of
// the Software, and to permit persons to whom the Software
// is furnished to do so, subject to the following
// conditions:
//
// The above copyright notice and this permission notice
// shall be included in all copies or substantial portions
// of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF
// ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED
// TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
// PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT
// SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
// CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR
// IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
// DEALINGS IN THE SOFTWARE.

//! Core traits and structs of the `jsonrpsee` library.
//!
//! > **Note**: This document mostly covers the core internal concepts of the `jsonrpsee` library.
//! >           See instead the documentation of `jsonrpsee` for how to use the library if you
//! >           just want to use it.
//!
//! jsonrpsee-core has five main concepts:
//!
//! - The [`common`] module contains all the primitive types of the JSON-RPC protocol, and
//!   utilities to convert between them and JSON.
//! - The [`TransportClient`] and [`TransportServer`] traits are implemented on structs that allow performing
//!   low-level communication with respectively a server or a client. These are the traits that
//!   you must implement if you are writing a custom transport (similar to HTTP, WebSockets,
//!   IPC, etc.).
//! - The [`RawClient`] and [`RawServer`] structs wrap around respectively a [`TransportClient`] or a
//!   [`TransportServer`] and allow correctly associating requests with responses and managing pub-sub
//!   subscriptions.
//!
//! In order to start a client or a server, first create a struct that implements respectively
//! [`TransportClient`] or [`TransportServer`], then wrap a [`RawClient`] or a [`RawServer`] around them.

pub use crate::core::client::raw::TransportClient;
pub use crate::core::client::{RawClient, RawClientError, RawClientEvent, RawClientRequestId};
pub use crate::core::local::local_transport;
pub use crate::core::server::raw::{TransportServer, TransportServerEvent};
pub use crate::core::server::{
    RawServer, RawServerEvent, RawServerRequestId, RawServerSubscriptionId,
};

pub mod client;
pub mod common;
pub mod local;
pub mod server;
