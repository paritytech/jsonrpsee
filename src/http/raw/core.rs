// Copyright (c) 2019 Parity Technologies Limited
//
// Permission is hereby granted, free of charge, to any
// person obtaining a copy of this software and associated
// documentation files (the "Software"), to deal in the
// Software without restriction, including without
// limitation the rights to use, copy, modify, merge,
// publish, distribute, sublicense, and/or sell copies of
// the Software, and to permit persons to whom the Software
// is furnished to do so, subject to the following
// conditions:
//
// The above copyright notice and this permission notice
// shall be included in all copies or substantial portions
// of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF
// ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED
// TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
// PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT
// SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
// CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR
// IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
// DEALINGS IN THE SOFTWARE.

use crate::http::transport::{HttpTransportServer, TransportServerEvent};
use crate::types::jsonrpc;
use crate::types::jsonrpc::wrapped::{batches, Notification, Params};

use core::{fmt, hash::Hash};

pub type RequestId = u64;

/// Wraps around a "raw server" and adds capabilities.
///
/// See the module-level documentation for more information.
pub struct RawServer {
	/// Internal "raw" server.
	raw: HttpTransportServer,

	/// List of requests that are in the progress of being answered. Each batch is associated with
	/// the raw request ID, or with `None` if this raw request has been closed.
	///
	/// See the documentation of [`BatchesState`][batches::BatchesState] for more information.
	batches: batches::BatchesState<Option<RequestId>>,
}

/// Identifier of a request within a `RawServer`.
#[derive(Debug, Copy, Clone, PartialEq, Eq)]
pub struct RawServerRequestId {
	inner: batches::BatchesElemId,
}

/// Identifier of a subscription within a [`RawServer`](crate::server::RawServer).
#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash)]
pub struct RawServerSubscriptionId([u8; 32]);

/// Event generated by a [`RawServer`](crate::server::RawServer).
///
/// > **Note**: Holds a borrow of the `RawServer`. Therefore, must be dropped before the `RawServer` can
/// >           be dropped.
#[derive(Debug)]
pub enum RawServerEvent<'a> {
	/// Request is a notification.
	Notification(Notification),

	/// Request is a method call.
	Request(RawServerRequest<'a>),
}

/// Request received by a [`RawServer`](crate::raw::RawServer).
pub struct RawServerRequest<'a> {
	/// Reference to the request within `self.batches`.
	inner: batches::BatchesElem<'a, Option<RequestId>>,

	/// Reference to the corresponding field in `RawServer`.
	raw: &'a mut HttpTransportServer,
}

impl RawServer {
	/// Starts a [`RawServer`](crate::raw::RawServer) using the given raw server internally.
	pub fn new(raw: HttpTransportServer) -> RawServer {
		RawServer { raw, batches: batches::BatchesState::new() }
	}
}

impl RawServer {
	/// Returns a `Future` resolving to the next event that this server generates.
	pub async fn next_event<'a>(&'a mut self) -> RawServerEvent<'a> {
		let request_id = loop {
			match self.batches.next_event() {
				None => {}
				Some(batches::BatchesEvent::Notification { notification, .. }) => {
					return RawServerEvent::Notification(notification)
				}
				Some(batches::BatchesEvent::Request(inner)) => {
					break RawServerRequestId { inner: inner.id() };
				}
				Some(batches::BatchesEvent::ReadyToSend { response, user_param: Some(raw_request_id) }) => {
					let _ = self.raw.finish(&raw_request_id, Some(&response)).await;
					continue;
				}
				Some(batches::BatchesEvent::ReadyToSend { response: _, user_param: None }) => {
					// This situation happens if the connection has been closed by the client.
					// Clients who close their connection.
					continue;
				}
			};

			match self.raw.next_request().await {
				TransportServerEvent::Request { id, request } => self.batches.inject(request, Some(id)),
				TransportServerEvent::Closed(raw_id) => {
					// The client has a closed their connection. We eliminate all traces of the
					// raw request ID from our state.
					// TODO: this has an O(n) complexity; make sure that this is not attackable
					for ud in self.batches.batches() {
						if ud.as_ref() == Some(&raw_id) {
							*ud = None;
						}
					}
				}
			};
		};
		RawServerEvent::Request(self.request_by_id(&request_id).unwrap())
	}

	/// Returns a request previously returned by [`next_event`](crate::raw::RawServer::next_event)
	/// by its id.
	///
	/// Note that previous notifications don't have an ID and can't be accessed with this method.
	///
	/// Returns `None` if the request ID is invalid or if the request has already been answered in
	/// the past.
	pub fn request_by_id<'a>(&'a mut self, id: &RawServerRequestId) -> Option<RawServerRequest<'a>> {
		Some(RawServerRequest { inner: self.batches.request_by_id(id.inner)?, raw: &mut self.raw })
	}
}

impl From<HttpTransportServer> for RawServer {
	fn from(inner: HttpTransportServer) -> Self {
		RawServer::new(inner)
	}
}

impl<'a> RawServerRequest<'a> {
	/// Returns the id of the request.
	///
	/// If this request object is dropped, you can retrieve it again later by calling
	/// [`request_by_id`](crate::raw::RawServer::request_by_id).
	pub fn id(&self) -> RawServerRequestId {
		RawServerRequestId { inner: self.inner.id() }
	}

	/// Returns the id that the client sent out.
	// TODO: can return None, which is wrong
	pub fn request_id(&self) -> &jsonrpc::Id {
		self.inner.request_id()
	}

	/// Returns the method of this request.
	pub fn method(&self) -> &str {
		self.inner.method()
	}

	/// Returns the parameters of the request, as a `jsonrpc::Params`.
	pub fn params(&self) -> Params {
		self.inner.params()
	}
}

impl<'a> RawServerRequest<'a> {
	/// Send back a response.
	///
	/// If this request is part of a batch:
	///
	/// - If all requests of the batch have been responded to, then the response is actively
	///   sent out.
	/// - Otherwise, this response is buffered.
	///
	/// > **Note**: This method is implemented in a way that doesn't wait for long to send the
	/// >           response. While calling this method will block your entire server, it
	/// >           should only block it for a short amount of time. See also [the equivalent
	/// >           method](crate::transport::TransportServer::finish) on the
	/// >           [`TransportServer`](crate::transport::TransportServer) trait.
	///
	pub fn respond(self, response: Result<jsonrpc::JsonValue, jsonrpc::Error>) {
		self.inner.set_response(response);
		//unimplemented!();
		// TODO: actually send out response?
	}
}

impl<'a> fmt::Debug for RawServerRequest<'a> {
	fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
		f.debug_struct("RawServerRequest")
			.field("request_id", &self.request_id())
			.field("method", &self.method())
			.field("params", &self.params())
			.finish()
	}
}
