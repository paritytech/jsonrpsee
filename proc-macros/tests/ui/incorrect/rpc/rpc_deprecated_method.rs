//! Test that calling a deprecated method will generate warnings at compile-time.

// Treat warnings as errors to fail the build.
#![deny(warnings)]

use std::net::SocketAddr;

use jsonrpsee::proc_macros::rpc;
use jsonrpsee::core::{async_trait, RpcResult};
use jsonrpsee::ws_client::*;
use jsonrpsee::ws_server::WsServerBuilder;

#[rpc(client, server)]
pub trait Deprecated {
	// Deprecated method that is called by the client.
	#[deprecated(since = "0.5.0", note = "please use `new_method` instead")]
	#[method(name = "foo")]
	async fn async_method(&self) -> RpcResult<u8>;

	// Deprecated methods that are not called should not generate warnings.
	#[deprecated(since = "0.5.0", note = "please use `new_method` instead")]
	#[method(name = "foo_unused")]
	async fn async_method_unused(&self) -> RpcResult<u8>;

	// If the method is not marked as deprecated, should not generate warnings.
	#[method(name = "bar")]
	fn sync_method(&self) -> RpcResult<u8>;
}

pub struct DeprecatedServerImpl;

#[async_trait]
impl DeprecatedServer for DeprecatedServerImpl {
	async fn async_method(&self) -> RpcResult<u8> {
		Ok(16u8)
	}

	async fn async_method_unused(&self) -> RpcResult<u8> {
		Ok(32u8)
	}

	fn sync_method(&self) -> RpcResult<u8> {
		Ok(64u8)
	}
}

pub async fn websocket_server() -> SocketAddr {
	let server = WsServerBuilder::default().build("127.0.0.1:0").await.unwrap();
	let addr = server.local_addr().unwrap();

	server.start(DeprecatedServerImpl.into_rpc()).unwrap();

	addr
}

#[tokio::main]
async fn main() {
	let server_addr = websocket_server().await;
	let server_url = format!("ws://{}", server_addr);
	let client = WsClientBuilder::default().build(&server_url).await.unwrap();

	// Calling this method should generate an warning.
	assert_eq!(client.async_method().await.unwrap(), 16);
	// Note: `async_method_unused` is not called, and should not generate warnings.
	assert_eq!(client.sync_method().await.unwrap(), 64);
}
