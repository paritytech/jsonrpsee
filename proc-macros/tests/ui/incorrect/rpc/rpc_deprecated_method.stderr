error: use of deprecated method `DeprecatedClient::async_method`: please use `new_method` instead
  --> $DIR/rpc_deprecated_method.rs:63:20
   |
63 |     assert_eq!(client.async_method().await.unwrap(), 16);
   |                       ^^^^^^^^^^^^
   |
note: the lint level is defined here
  --> $DIR/rpc_deprecated_method.rs:4:9
   |
4  | #![deny(warnings)]
   |         ^^^^^^^^
   = note: `#[deny(deprecated)]` implied by `#[deny(warnings)]`
